File: apps/android/app/src/main/java/com/mollotov/browser/ai/AIHandler.kt
package com.mollotov.browser.ai

import android.content.Context
import android.os.SystemClock
import com.mollotov.browser.network.Router
import com.mollotov.browser.network.errorResponse
import com.mollotov.browser.network.successResponse
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.double
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject

private val aiJson = Json { ignoreUnknownKeys = true; isLenient = true }

class AIHandler(private val appContext: Context) {
    private val platformEngine by lazy { PlatformAIEngine(appContext) }
    private val recorder by lazy { AudioRecorder(appContext) }

    fun register(router: Router) {
        router.register("ai-status") { aiStatus() }
        router.register("ai-load") { aiLoad(it) }
        router.register("ai-unload") { aiUnload() }
        router.register("ai-infer") { aiInfer(it) }
        router.register("ai-record") { aiRecord(it) }
    }

    private fun aiStatus(): Map<String, Any?> {
        val backend = currentBackend()
        val model = currentModel()
        val data = linkedMapOf<String, Any?>(
            "loaded" to (model != null),
            "backend" to backend,
            "capabilities" to capabilitiesFor(backend, model),
        )
        if (model != null) {
            data["model"] = model
        }
        if (backend == AIState.OLLAMA_BACKEND) {
            data["ollamaEndpoint"] = AIState.ollamaEndpoint
        }
        return successResponse(data)
    }

    private suspend fun aiLoad(body: Map<String, Any?>): Map<String, Any?> {
        val requestedModel = (body["model"] as? String)?.trim().orEmpty()
        val start = SystemClock.elapsedRealtime()

        if (requestedModel.isEmpty() || requestedModel == AIState.PLATFORM_MODEL_ID) {
            if (!AIState.isAvailable) {
                return errorResponse(
                    "PLATFORM_AI_UNAVAILABLE",
                    "Platform AI is not available on this device",
                )
            }

            AIState.backend = AIState.PLATFORM_BACKEND
            AIState.activeModel = null

            return successResponse(
                mapOf(
                    "model" to AIState.PLATFORM_MODEL_ID,
                    "backend" to AIState.PLATFORM_BACKEND,
                    "loadTimeMs" to (SystemClock.elapsedRealtime() - start),
                ),
            )
        }

        if (!requestedModel.startsWith("ollama:")) {
            return errorResponse(
                "MODEL_NOT_FOUND",
                "Android supports the platform backend or ollama: model IDs",
            )
        }

        val endpoint = normalizeEndpoint(
            (body["ollamaEndpoint"] as? String)?.trim().takeUnless { it.isNullOrEmpty() }
                ?: AIState.ollamaEndpoint
                ?: AIState.DEFAULT_OLLAMA_ENDPOINT,
        )
        val ollamaModel = requestedModel.removePrefix("ollama:")

        val installedModels = try {
            fetchOllamaModels(endpoint)
        } catch (_: IOException) {
            return errorResponse("OLLAMA_NOT_AVAILABLE", "Ollama is not running at $endpoint")
        } catch (e: Exception) {
            return errorResponse("AI_INFERENCE_FAILED", e.message ?: "Failed to probe Ollama")
        }

        if (ollamaModel !in installedModels) {
            return errorResponse(
                "OLLAMA_MODEL_NOT_FOUND",
                "Ollama model '$ollamaModel' is not installed at $endpoint",
            )
        }

        AIState.backend = AIState.OLLAMA_BACKEND
        AIState.activeModel = ollamaModel
        AIState.ollamaEndpoint = endpoint

        return successResponse(
            mapOf(
                "model" to ollamaModel,
                "backend" to AIState.OLLAMA_BACKEND,
                "loadTimeMs" to (SystemClock.elapsedRealtime() - start),
            ),
        )
    }

    private fun aiUnload(): Map<String, Any?> {
        AIState.backend = AIState.PLATFORM_BACKEND
        AIState.activeModel = null

        return successResponse(
            mapOf(
                "backend" to currentBackend(),
                "model" to currentModel(),
            ),
        )
    }

    private suspend fun aiInfer(body: Map<String, Any?>): Map<String, Any?> {
        return when (currentBackend()) {
            AIState.OLLAMA_BACKEND -> inferWithOllama(body)
            AIState.PLATFORM_BACKEND -> inferWithPlatform(body)
            else -> errorResponse("NO_MODEL_LOADED", "Load a model first with ai-load")
        }
    }

    private fun aiRecord(body: Map<String, Any?>): Map<String, Any?> {
        val action = (body["action"] as? String)?.trim().orEmpty().ifEmpty { "status" }
        return try {
            when (action) {
                "start" -> {
                    recorder.start()
                    successResponse(mapOf("recording" to true, "elapsedMs" to 0))
                }
                "stop" -> {
                    val result = recorder.stop()
                    successResponse(
                        mapOf(
                            "recording" to false,
                            "audio" to android.util.Base64.encodeToString(result.audio, android.util.Base64.NO_WRAP),
                            "durationMs" to result.durationMs,
                        ),
                    )
                }
                "status" -> {
                    successResponse(
                        mapOf(
                            "recording" to recorder.isRecording,
                            "elapsedMs" to recorder.elapsedMs,
                        ),
                    )
                }
                else -> errorResponse("INVALID_PARAM", "action must be start, stop, or status")
            }
        } catch (e: SecurityException) {
            errorResponse("MIC_PERMISSION_DENIED", "Microphone permission not granted")
        } catch (e: IllegalStateException) {
            val code = when {
                e.message?.contains("ALREADY_ACTIVE") == true -> "RECORDING_ALREADY_ACTIVE"
                e.message?.contains("NO_RECORDING") == true -> "NO_RECORDING_ACTIVE"
                else -> "RECORDING_FAILED"
            }
            errorResponse(code, e.message ?: "Recording failed")
        }
    }

    private suspend fun inferWithPlatform(body: Map<String, Any?>): Map<String, Any?> {
        if (!AIState.isAvailable) {
            return errorResponse(
                "PLATFORM_AI_UNAVAILABLE",
                "Platform AI is not available on this device",
            )
        }
        if (body["image"] != null || body["images"] != null) {
            return errorResponse("VISION_NOT_SUPPORTED", "Platform AI on Android is text-only for now")
        }
        if (body["audio"] != null) {
            return errorResponse(
                "AUDIO_NOT_SUPPORTED",
                "Platform AI on Android does not accept audio input yet",
            )
        }

        val prompt = buildPrompt(body)
            ?: return errorResponse("MISSING_PARAM", "prompt is required")
        val start = SystemClock.elapsedRealtime()

        return try {
            val response = platformEngine.infer(prompt)
            successResponse(
                mapOf(
                    "response" to response,
                    "tokensUsed" to estimateTokens(response),
                    "inferenceTimeMs" to (SystemClock.elapsedRealtime() - start),
                ),
            )
        } catch (e: UnsupportedOperationException) {
            errorResponse("PLATFORM_AI_NOT_WIRED", e.message ?: "Platform AI is not yet wired")
        } catch (e: Exception) {
            errorResponse("AI_INFERENCE_FAILED", e.message ?: "Platform AI inference failed")
        }
    }

    private suspend fun inferWithOllama(body: Map<String, Any?>): Map<String, Any?> {
        val endpoint = AIState.ollamaEndpoint
            ?: return errorResponse("OLLAMA_NOT_AVAILABLE", "Ollama endpoint is not configured")
        val model = AIState.activeModel
            ?: return errorResponse("NO_MODEL_LOADED", "Load a model first with ai-load")

        if (body["audio"] != null) {
            return errorResponse(
                "AUDIO_NOT_SUPPORTED",
                "Android does not proxy audio requests to Ollama yet",
            )
        }

        val prompt = buildPrompt(body)
            ?: return errorResponse("MISSING_PARAM", "prompt is required")
        val messages = parseMessages(body["messages"])
        val images = extractImages(body)
        val start = SystemClock.elapsedRealtime()

        return try {
            val isChat = messages.isNotEmpty()
            val response = if (isChat) {
                postJson(
                    url = "$endpoint/api/chat",
                    payload = buildChatRequest(model, messages, prompt, body),
                )
            } else {
                postJson(
                    url = "$endpoint/api/generate",
                    payload = buildGenerateRequest(model, prompt, images, body),
                )
            }

            val responseText = if (isChat) {
                val message = response["message"] as? Map<*, *>
                message?.get("content") as? String
            } else {
                response["response"] as? String
            }?.trim().orEmpty()

            val tokensUsed = (response["eval_count"] as? Number)?.toInt() ?: estimateTokens(responseText)
            successResponse(
                mapOf(
                    "response" to responseText,
                    "tokensUsed" to tokensUsed,
                    "inferenceTimeMs" to (SystemClock.elapsedRealtime() - start),
                ),
            )
        } catch (_: IOException) {
            errorResponse("OLLAMA_DISCONNECTED", "Lost connection to Ollama during inference")
        } catch (e: Exception) {
            errorResponse("AI_INFERENCE_FAILED", e.message ?: "Ollama inference failed")
        }
    }

    private fun currentBackend(): String {
        return AIState.backend
    }

    private fun currentModel(): String? {
        return when (currentBackend()) {
            AIState.OLLAMA_BACKEND -> AIState.activeModel
            AIState.PLATFORM_BACKEND -> if (AIState.isAvailable) AIState.PLATFORM_MODEL_ID else null
            else -> null
        }
    }

    private fun capabilitiesFor(backend: String, model: String?): List<String> {
        return when (backend) {
            AIState.OLLAMA_BACKEND -> if (looksVisionCapable(model)) listOf("text", "vision") else listOf("text")
            AIState.PLATFORM_BACKEND -> if (AIState.isAvailable) listOf("text") else emptyList()
            else -> emptyList()
        }
    }

    private fun looksVisionCapable(model: String?): Boolean {
        if (model == null) return false
        val lowercase = model.lowercase()
        return listOf("llava", "vision", "moondream", "minicpm-v").any { lowercase.contains(it) }
    }

    private fun buildPrompt(body: Map<String, Any?>): String? {
        val prompt = (body["prompt"] as? String)?.trim().orEmpty()
        val text = (body["text"] as? String)?.trim().orEmpty()

        if (prompt.isEmpty() && text.isEmpty()) return null
        if (prompt.isEmpty()) return text
        if (text.isEmpty()) return prompt
        return "$prompt\n\n$text"
    }

    private fun extractImages(body: Map<String, Any?>): List<String> {
        val explicitImages = (body["images"] as? List<*>)
            ?.mapNotNull { it as? String }
            ?.filter { it.isNotBlank() }
            .orEmpty()
        if (explicitImages.isNotEmpty()) {
            return explicitImages
        }

        val singleImage = (body["image"] as? String)?.takeIf { it.isNotBlank() }
        return if (singleImage != null) listOf(singleImage) else emptyList()
    }

    private fun parseMessages(value: Any?): List<Map<String, String>> {
        val rawMessages = value as? List<*> ?: return emptyList()
        return rawMessages.mapNotNull { item ->
            val map = item as? Map<*, *> ?: return@mapNotNull null
            val role = map["role"] as? String ?: return@mapNotNull null
            val content = map["content"] as? String ?: return@mapNotNull null
            mapOf("role" to role, "content" to content)
        }
    }

    private fun buildGenerateRequest(
        model: String,
        prompt: String,
        images: List<String>,
        body: Map<String, Any?>,
    ): Map<String, Any?> {
        val request = linkedMapOf<String, Any?>(
            "model" to model,
            "prompt" to prompt,
            "stream" to false,
        )
        if (images.isNotEmpty()) {
            request["images"] = images
        }
        buildOllamaOptions(body)?.let { request["options"] = it }
        return request
    }

    private fun buildChatRequest(
        model: String,
        messages: List<Map<String, String>>,
        prompt: String,
        body: Map<String, Any?>,
    ): Map<String, Any?> {
        val requestMessages = messages.toMutableList()
        requestMessages.add(mapOf("role" to "user", "content" to prompt))

        val request = linkedMapOf<String, Any?>(
            "model" to model,
            "messages" to requestMessages,
            "stream" to false,
        )
        buildOllamaOptions(body)?.let { request["options"] = it }
        return request
    }

    private fun buildOllamaOptions(body: Map<String, Any?>): Map<String, Any?>? {
        val options = linkedMapOf<String, Any?>()

        val maxTokens = body["maxTokens"] as? Number
        if (maxTokens != null) {
            options["num_predict"] = maxTokens.toInt()
        }

        val temperature = body["temperature"] as? Number
        if (temperature != null) {
            options["temperature"] = temperature.toDouble()
        }

        return options.takeIf { it.isNotEmpty() }
    }

    private suspend fun fetchOllamaModels(endpoint: String): List<String> {
        val response = getJson("$endpoint/api/tags")
        val models = response["models"] as? List<*> ?: return emptyList()
        return models.mapNotNull { item ->
            val model = item as? Map<*, *> ?: return@mapNotNull null
            model["name"] as? String
        }
    }

    private suspend fun getJson(url: String): Map<String, Any?> = withContext(Dispatchers.IO) {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 3_000
            readTimeout = 10_000
            doInput = true
        }

        try {
            val body = readConnectionBody(connection)
            if (connection.responseCode !in 200..299) {
                throw IOException("HTTP ${connection.responseCode}: $body")
            }
            parseJsonObject(body)
        } finally {
            connection.disconnect()
        }
    }

    private suspend fun postJson(url: String, payload: Map<String, Any?>): Map<String, Any?> =
        withContext(Dispatchers.IO) {
            val connection = (URL(url).openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 3_000
                readTimeout = 30_000
                doInput = true
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Accept", "application/json")
            }

            try {
                connection.outputStream.bufferedWriter().use { writer ->
                    writer.write(mapToJsonString(payload))
                }

                val body = readConnectionBody(connection)
                if (connection.responseCode !in 200..299) {
                    throw IOException("HTTP ${connection.responseCode}: $body")
                }
                parseJsonObject(body)
            } finally {
                connection.disconnect()
            }
        }

    private fun readConnectionBody(connection: HttpURLConnection): String {
        val stream = if (connection.responseCode in 200..299) {
            connection.inputStream
        } else {
            connection.errorStream
        } ?: return ""

        return stream.bufferedReader().use { it.readText() }
    }

    private fun mapToJsonString(map: Map<String, Any?>): String {
        return aiJson.encodeToString(JsonElement.serializer(), mapToJsonObject(map))
    }

    private fun mapToJsonObject(map: Map<String, Any?>): JsonObject {
        return JsonObject(map.entries.associate { (key, value) -> key to anyToJsonElement(value) })
    }

    @Suppress("UNCHECKED_CAST")
    private fun anyToJsonElement(value: Any?): JsonElement = when (value) {
        null -> JsonNull
        is Boolean -> JsonPrimitive(value)
        is Int -> JsonPrimitive(value)
        is Long -> JsonPrimitive(value)
        is Double -> JsonPrimitive(value)
        is Float -> JsonPrimitive(value)
        is String -> JsonPrimitive(value)
        is Map<*, *> -> mapToJsonObject(value as Map<String, Any?>)
        is List<*> -> JsonArray(value.map { anyToJsonElement(it) })
        else -> JsonPrimitive(value.toString())
    }

    private fun parseJsonObject(text: String): Map<String, Any?> {
        if (text.isBlank()) return emptyMap()
        val element = aiJson.parseToJsonElement(text)
        return jsonObjectToMap(element.jsonObject)
    }

    private fun jsonObjectToMap(obj: JsonObject): Map<String, Any?> {
        return obj.entries.associate { (key, value) -> key to jsonElementToAny(value) }
    }

    private fun jsonElementToAny(element: JsonElement): Any? = when (element) {
        is JsonNull -> null
        is JsonPrimitive -> when {
            element.isString -> element.content
            element.booleanOrNull != null -> element.boolean
            element.intOrNull != null -> element.int
            element.doubleOrNull != null -> element.double
            else -> element.content
        }
        is JsonObject -> jsonObjectToMap(element)
        is JsonArray -> element.map { jsonElementToAny(it) }
    }

    private fun estimateTokens(text: String): Int {
        return (text.length / 4.0).toInt().coerceAtLeast(1)
    }

    private fun normalizeEndpoint(endpoint: String): String {
        return endpoint.trim().trimEnd('/')
    }
}

File: apps/ios/Mollotov/Handlers/AIHandler.swift
import AVFoundation
import Foundation

struct AIHandler {
    let context: HandlerContext

    private let platformEngine = PlatformAIEngine()
    private static let ollamaPrefix = "ollama:"
    private static let recorder = StubAudioRecorder()

    func register(on router: Router) {
        router.register("ai-status") { _ in await status() }
        router.register("ai-load") { body in await load(body) }
        router.register("ai-unload") { _ in await unload() }
        router.register("ai-infer") { body in await infer(body) }
        router.register("ai-record") { body in await record(body) }
    }

    private func status() async -> [String: Any] {
        await MainActor.run {
            let state = AIState.shared
            var response = successResponse([
                "loaded": state.isLoaded,
                "backend": state.backend,
                "capabilities": state.capabilities,
            ])

            if state.backend == "ollama", let model = state.activeModel {
                response["model"] = model
                response["ollamaEndpoint"] = state.ollamaEndpoint
            }

            return response
        }
    }

    private func load(_ body: [String: Any]) async -> [String: Any] {
        let start = CFAbsoluteTimeGetCurrent()
        let model = (body["model"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)

        if let model, model.hasPrefix(Self.ollamaPrefix) {
            let ollamaModel = String(model.dropFirst(Self.ollamaPrefix.count))
            guard !ollamaModel.isEmpty else {
                return errorResponse(code: "INVALID_PARAM", message: "Ollama model id cannot be empty")
            }

            let endpoint = normalizedOllamaEndpoint(body["ollamaEndpoint"] as? String)
            guard URL(string: endpoint) != nil else {
                return errorResponse(code: "INVALID_PARAM", message: "ollamaEndpoint must be a valid URL")
            }

            do {
                try await validateOllamaModel(model: ollamaModel, endpoint: endpoint)
                await MainActor.run {
                    AIState.shared.activateOllama(model: ollamaModel, endpoint: endpoint)
                }
                return successResponse([
                    "model": ollamaModel,
                    "backend": "ollama",
                    "ollamaEndpoint": endpoint,
                    "loadTimeMs": elapsedMs(since: start),
                ])
            } catch let error as AIHandlerError {
                return error.response
            } catch {
                return errorResponse(code: "OLLAMA_NOT_AVAILABLE", message: "Failed to reach Ollama at \(endpoint)")
            }
        }

        if body["path"] != nil {
            return errorResponse(
                code: "PLATFORM_NOT_SUPPORTED",
                message: "iOS does not support loading GGUF files directly"
            )
        }

        if let model, !model.isEmpty, model != "platform" {
            return errorResponse(
                code: "PLATFORM_NOT_SUPPORTED",
                message: "iOS only supports the platform backend or ollama: prefixed models"
            )
        }

        let payload = await MainActor.run { () -> [String: Any] in
            let state = AIState.shared
            state.activatePlatform()
            return [
                "loaded": state.isLoaded,
                "backend": state.backend,
                "capabilities": state.capabilities,
                "loadTimeMs": elapsedMs(since: start),
            ]
        }
        return successResponse(payload)
    }

    private func unload() async -> [String: Any] {
        await MainActor.run {
            AIState.shared.activatePlatform()
            return successResponse()
        }
    }

    private func infer(_ body: [String: Any]) async -> [String: Any] {
        let stateSnapshot = await MainActor.run { () -> (backend: String, model: String?, endpoint: String, available: Bool) in
            let state = AIState.shared
            return (state.backend, state.activeModel, state.ollamaEndpoint, state.isAvailable)
        }

        switch stateSnapshot.backend {
        case "ollama":
            guard let model = stateSnapshot.model, !model.isEmpty else {
                return errorResponse(code: "NO_MODEL_LOADED", message: "Load a model first with ai-load")
            }
            return await inferWithOllama(body, model: model, endpoint: stateSnapshot.endpoint)
        case "platform":
            guard stateSnapshot.available else {
                return errorResponse(code: "PLATFORM_AI_UNAVAILABLE", message: "Platform AI is unavailable on this device")
            }
            return await inferWithPlatform(body)
        default:
            return errorResponse(code: "INVALID_STATE", message: "Unknown AI backend: \(stateSnapshot.backend)")
        }
    }

    private func inferWithPlatform(_ body: [String: Any]) async -> [String: Any] {
        if body["audio"] != nil {
            return errorResponse(
                code: "AUDIO_NOT_SUPPORTED",
                message: "Platform AI currently supports text-only inference on iOS"
            )
        }

        if body["image"] != nil || body["images"] != nil || (body["context"] as? String) == "screenshot" {
            return errorResponse(
                code: "VISION_NOT_SUPPORTED",
                message: "Platform AI currently supports text-only inference on iOS"
            )
        }

        if body["messages"] != nil {
            return errorResponse(
                code: "INVALID_PARAM",
                message: "messages is only supported when the Ollama backend is active"
            )
        }

        let prompt = composePrompt(from: body)
        guard !prompt.isEmpty else {
            return errorResponse(code: "MISSING_PARAM", message: "prompt is required")
        }

        let start = CFAbsoluteTimeGetCurrent()

        do {
            let response = try await platformEngine.infer(prompt: prompt)
            return successResponse([
                "response": response,
                "tokensUsed": 0,
                "inferenceTimeMs": elapsedMs(since: start),
            ])
        } catch {
            return errorResponse(
                code: "PLATFORM_AI_UNAVAILABLE",
                message: error.localizedDescription.isEmpty ? "Platform AI is not yet wired on iOS" : error.localizedDescription
            )
        }
    }

    private func inferWithOllama(_ body: [String: Any], model: String, endpoint: String) async -> [String: Any] {
        if body["audio"] != nil {
            return errorResponse(
                code: "AUDIO_NOT_SUPPORTED",
                message: "Remote Ollama inference does not accept raw audio in the iOS bridge"
            )
        }

        let start = CFAbsoluteTimeGetCurrent()

        do {
            if let messages = body["messages"] as? [[String: Any]] {
                let payload = try buildOllamaChatPayload(body: body, messages: messages, model: model)
                let result = try await postOllama(path: "/api/chat", payload: payload, endpoint: endpoint)
                let message = result["message"] as? [String: Any]
                let content = message?["content"] as? String ?? ""

                return successResponse([
                    "response": content,
                    "tokensUsed": result["eval_count"] as? Int ?? 0,
                    "inferenceTimeMs": elapsedMs(since: start),
                ])
            }

            let payload = try buildOllamaGeneratePayload(body: body, model: model)
            let result = try await postOllama(path: "/api/generate", payload: payload, endpoint: endpoint)
            let response = result["response"] as? String ?? ""

            return successResponse([
                "response": response,
                "tokensUsed": result["eval_count"] as? Int ?? 0,
                "inferenceTimeMs": elapsedMs(since: start),
            ])
        } catch let error as AIHandlerError {
            return error.response
        } catch {
            return errorResponse(code: "OLLAMA_DISCONNECTED", message: "Lost connection to Ollama during inference")
        }
    }

    private func record(_ body: [String: Any]) async -> [String: Any] {
        let action = (body["action"] as? String ?? "status").lowercased()

        do {
            switch action {
            case "start":
                try Self.recorder.start()
                return successResponse(["recording": true, "elapsedMs": 0])
            case "stop":
                let result = try Self.recorder.stop()
                return successResponse([
                    "recording": false,
                    "audio": result.audio.base64EncodedString(),
                    "durationMs": result.durationMs,
                ])
            case "status":
                let snapshot: [String: Any] = [
                    "recording": Self.recorder.isRecording,
                    "elapsedMs": Self.recorder.elapsedMs,
                ]
                return successResponse(snapshot)
            default:
                return errorResponse(code: "INVALID_PARAM", message: "action must be start, stop, or status")
            }
        } catch let error as RecorderError {
            return error.response
        } catch {
            return errorResponse(code: "RECORDING_FAILED", message: "Audio recording is unavailable")
        }
    }

    private func validateOllamaModel(model: String, endpoint: String) async throws {
        let result = try await postOllama(path: "/api/tags", payload: nil, endpoint: endpoint, method: "GET")
        let models = result["models"] as? [[String: Any]] ?? []
        let names = Set(models.compactMap { $0["name"] as? String })

        guard names.contains(model) else {
            throw AIHandlerError(
                code: "OLLAMA_MODEL_NOT_FOUND",
                message: "Ollama model '\(model)' is not installed at \(endpoint)"
            )
        }
    }

    private func buildOllamaGeneratePayload(body: [String: Any], model: String) throws -> [String: Any] {
        let prompt = composePrompt(from: body)
        guard !prompt.isEmpty else {
            throw AIHandlerError(code: "MISSING_PARAM", message: "prompt is required")
        }

        var payload: [String: Any] = [
            "model": model,
            "prompt": prompt,
            "stream": false,
        ]

        if let maxTokens = body["maxTokens"] as? Int {
            payload["options"] = ["num_predict": maxTokens]
        }

        let images = normalizedImages(from: body)
        if !images.isEmpty {
            payload["images"] = images
        }

        return payload
    }

    private func buildOllamaChatPayload(
        body: [String: Any],
        messages: [[String: Any]],
        model: String
    ) throws -> [String: Any] {
        var normalizedMessages = messages

        if let prompt = nonEmptyString(body["prompt"]) {
            normalizedMessages.append([
                "role": "user",
                "content": prompt,
            ])
        }

        guard !normalizedMessages.isEmpty else {
            throw AIHandlerError(code: "MISSING_PARAM", message: "messages or prompt is required")
        }

        return [
            "model": model,
            "messages": normalizedMessages,
            "stream": false,
        ]
    }

    private func postOllama(
        path: String,
        payload: [String: Any]?,
        endpoint: String,
        method: String = "POST"
    ) async throws -> [String: Any] {
        guard let baseURL = URL(string: endpoint) else {
            throw AIHandlerError(code: "INVALID_PARAM", message: "ollamaEndpoint must be a valid URL")
        }

        let url = baseURL.appending(path: path.trimmingCharacters(in: CharacterSet(charactersIn: "/")))
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 30

        if let payload {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        }

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else {
                throw AIHandlerError(code: "OLLAMA_DISCONNECTED", message: "Invalid response from Ollama")
            }

            if !(200...299).contains(httpResponse.statusCode) {
                let message = String(data: data, encoding: .utf8) ?? "Ollama request failed"
                let code = httpResponse.statusCode == 404 ? "OLLAMA_MODEL_NOT_FOUND" : "OLLAMA_DISCONNECTED"
                throw AIHandlerError(code: code, message: message)
            }

            if data.isEmpty {
                return [:]
            }

            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw AIHandlerError(code: "OLLAMA_DISCONNECTED", message: "Ollama returned invalid JSON")
            }

            return json
        } catch let error as AIHandlerError {
            throw error
        } catch {
            throw AIHandlerError(code: "OLLAMA_DISCONNECTED", message: "Lost connection to Ollama during inference")
        }
    }

    private func composePrompt(from body: [String: Any]) -> String {
        let prompt = nonEmptyString(body["prompt"])
        let text = nonEmptyString(body["text"])

        switch (prompt, text) {
        case let (.some(prompt), .some(text)):
            return "\(prompt)\n\nContext:\n\(text)"
        case let (.some(prompt), nil):
            return prompt
        case let (nil, .some(text)):
            return text
        default:
            return ""
        }
    }

    private func normalizedImages(from body: [String: Any]) -> [String] {
        if let images = body["images"] as? [String] {
            return images.filter { !$0.isEmpty }
        }

        if let image = body["image"] as? String, !image.isEmpty {
            return [image]
        }

        return []
    }

    private func nonEmptyString(_ value: Any?) -> String? {
        guard let string = value as? String else { return nil }
        let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private func normalizedOllamaEndpoint(_ endpoint: String?) -> String {
        let trimmed = endpoint?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let base = trimmed.isEmpty ? AIState.defaultOllamaEndpoint : trimmed
        return base.hasSuffix("/") ? String(base.dropLast()) : base
    }

    private func elapsedMs(since start: CFAbsoluteTime) -> Int {
        Int((CFAbsoluteTimeGetCurrent() - start) * 1000)
    }
}

private struct AIHandlerError: Error {
    let code: String
    let message: String

    var response: [String: Any] {
        errorResponse(code: code, message: message)
    }
}

private final class StubAudioRecorder {
    struct Result {
        let audio: Data
        let durationMs: Int
    }

    private var engine: AVAudioEngine?
    private(set) var isRecording = false
    private var startedAt: Date?

    var elapsedMs: Int {
        guard let startedAt else { return 0 }
        return Int(Date().timeIntervalSince(startedAt) * 1000)
    }

    func start() throws {
        guard !isRecording else { throw RecorderError.alreadyRecording }
        let newEngine = AVAudioEngine()
        newEngine.prepare()
        engine = newEngine
        startedAt = Date()
        isRecording = true
    }

    func stop() throws -> Result {
        guard isRecording else { throw RecorderError.notRecording }
        let duration = elapsedMs
        engine?.stop()
        engine = nil
        startedAt = nil
        isRecording = false
        return Result(audio: Data(), durationMs: duration)
    }
}

private enum RecorderError: Error {
    case alreadyRecording
    case notRecording

    var response: [String: Any] {
        switch self {
        case .alreadyRecording:
            return errorResponse(code: "RECORDING_ALREADY_ACTIVE", message: "Recording is already active")
        case .notRecording:
            return errorResponse(code: "NO_RECORDING_ACTIVE", message: "No recording is active")
        }
    }
}
